hibernate-core/src/main/java/org/hibernate/query/sqm/sql/BaseSqmToSqlAstConverter.java

```diff
@@ -3216,6 +3216,18 @@ private TableGroup consumeAttributeJoin(
 
 		getFromClauseIndex().register( sqmJoin, joinedTableGroup );
 		registerPluralTableGroupParts( joinedTableGroup );
+
+		if ( !sqmJoin.hasTreats() && sqmJoin.getNodeType().getSqmPathType() instanceof EntityDomainType ) {
+			final EntityDomainType<?> entityDomainType = (EntityDomainType<?>) sqmJoin.getNodeType().getSqmPathType();
+			final String entityName = entityDomainType.getHibernateEntityName();
+			final EntityPersister entityDescriptor = domainModel.findEntityDescriptor( entityName );
+			if ( entityDescriptor != null && entityDescriptor.getSuperMappingType() != null ) {
+				// This is a non-treated join with a subtype entity, register a TREAT entity name use
+				// to filter only the correctly typed records
+				( (SqmJoin) sqmJoin ).treatAs( entityDomainType );
```

Comment:
I don't know if changing the SQM tree here is a good idea. Can you please register an entity name use instead?

Reply:
I have, but now we're getting some test failures specific to Derby (looks like these might be related to their optimizer, but the generated SQL queries themselves look fine to me).